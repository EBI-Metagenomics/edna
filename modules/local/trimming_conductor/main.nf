
process TRIMMING_CONDUCTOR {
    // Parse the outputs of the std and general primer searches and output
    // flags into "trimming_conductor_out.txt"
    tag "$meta.id"
    label 'very_light'
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        "https://depot.galaxyproject.org/singularity/mgnify-pipelines-toolkit:${params.mpt_version}":
        "biocontainers/mgnify-pipelines-toolkit:${params.mpt_version}" }"

    input:
    tuple val(meta), path(general_primer_flag), path(std_primer_flag)

    output:
    tuple val(meta), path("*trimming_conductor_out*"), emit: trimming_conductor_out
    path "versions.yml"                              , emit: versions

    script:
    """
    gen_fwd_flag="\$(sed '1q;d' ${general_primer_flag})"
    gen_rev_flag="\$(sed '2q;d' ${general_primer_flag})"

    std_fwd_primer="\$(grep 'F' ${std_primer_flag} || true)"
    std_rev_primer="\$(grep 'R' ${std_primer_flag} || true)"

    fwd_trim_flag='none'
    rev_trim_flag='none'

    if [[ ! -z \$std_fwd_primer ]]; then
        std_fwd_flag='1'
        fwd_trim_flag='std'
    else
        std_fwd_flag='0'
        if [[ \$gen_fwd_flag -eq '1' ]]; then
            fwd_trim_flag='auto'
        else
            fwd_trim_flag='none'
        fi
    fi

    if [[ ! -z \$std_rev_primer ]]; then
        std_rev_flag='1'
        rev_trim_flag='std'
    else
        std_rev_flag='0'
        if [[ \$gen_rev_flag -eq '1' ]]; then
            rev_trim_flag='auto'
        else
            rev_trim_flag='none'
        fi
    fi

    echo -en "\${fwd_trim_flag}\\n\${rev_trim_flag}" > ${meta.id}_trimming_conductor_out_${meta.var_region}.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mgnify-pipelines-toolkit: ${params.mpt_version}
    END_VERSIONS
    """
}
