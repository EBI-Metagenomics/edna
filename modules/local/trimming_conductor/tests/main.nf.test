nextflow_process {

    name "Test Process TRIMMING_CONDUCTOR"
    script "../main.nf"
    process "TRIMMING_CONDUCTOR"
    tag "modules"
    tag "trimming_conductor"

    test("Should take results of both general and std primer flagging processes and output a decision about if/where/how to trim primers, generating an output file containing the decision for each direction") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false, var_region:"16S-V3-V4" ], // meta map
                    file('${projectDir}/modules/local/trimming_conductor/tests/data/test_general_primer_out.txt', checkIfExists: true),
                    file('${projectDir}/modules/local/trimming_conductor/tests/data/test_std_primers.fasta', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.trimming_conductor_out).match("trimming_conductor_out") }
            )
        }

    }

}
