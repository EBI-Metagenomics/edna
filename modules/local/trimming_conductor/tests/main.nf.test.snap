{
    "trimming_conductor_out": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "var_region": "16S-V3-V4"
                    },
                    "test_trimming_conductor_out_16S-V3-V4.txt:md5,d0dd56ac864be40ec2c5b2bbcdfaeb52"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T16:46:46.164307134"
    }
}